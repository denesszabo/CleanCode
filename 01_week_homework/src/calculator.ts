export class Calculator {

    public Add(a: number, b: number): number {
        return a + b;
    }

    public Sub(a: number, b: number): number {
        return a - b;
    }

}
