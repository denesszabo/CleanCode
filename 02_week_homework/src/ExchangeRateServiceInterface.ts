/**
 * Defines the currency exchange rate service.
 */
export interface ExchangeRateServiceInterface {

    /**
     * Get the exchange rate between two currencies.
     *
     * @param fromCurrency string
     *   The currency to convert from.
     * @param toCurrency string
     *   The currency to convert to.
     *
     * @trows InvalidCurrencyError
     *   If the given currency is invalid or unknown
     */
    getExchangeRate(fromCurrency: string, toCurrency: string): number;
}
