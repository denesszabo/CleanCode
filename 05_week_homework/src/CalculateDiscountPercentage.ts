enum DISCOUNT_LEVELS {
    standard = 5,
    silver = 10,
    gold = 20,
    unknown = 0,
}
export class DiscountCalculator {

    public getDiscountPercentageForLevel(level: string): number {
        const k = level as keyof typeof DISCOUNT_LEVELS;
        return Object.keys(DISCOUNT_LEVELS).includes(level) ?
            DISCOUNT_LEVELS[k] :
            0;
    }
}
