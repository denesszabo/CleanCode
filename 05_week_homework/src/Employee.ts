export class Employee {
    private name: string;
    private salary: number;

    constructor(name: string, salary: number) {
        this.name = name;
        this.salary = salary;
    }

    public getName(): string {
        return this.name;
    }

    public setName(name: string): void {
        this.name = name;
    }

    public getSalary(): number {
        return this.salary;
    }

    public setSalary(salary: number): void {
        this.salary = salary;
    }

    calculateSalary(): number {
        return this.salary;
    }
}
