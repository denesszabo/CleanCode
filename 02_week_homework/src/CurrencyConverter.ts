import { ExchangeRateServiceInterface } from "./ExchangeRateServiceInterface";
import {InvalidInputAmountError} from "./error/InvalidInputAmountError";
import {InvalidCurrencyError} from "./error/InvalidCurrencyError";
import {NetworkError} from "./error/NetworkError";

export class CurrencyConverter {

    private readonly FIXED_AMOUNT = 100;
    private readonly EXCHANGE_RATE_ERROR = -1;

    constructor(private exchangeRateService: ExchangeRateServiceInterface) { }

    public Convert(amount: number, fromCurrency: string, toCurrency: string): number {
        this.validateAmount(amount);
        const exchangeRate = this.getExchangeRate(fromCurrency, toCurrency);

        this.validateExchangeRate(exchangeRate);
        return amount * exchangeRate;
    }

    /**
     * Conversion report builder for the give date range.
     *
     * @param fromCurrency
     * @param toCurrency
     * @param startDate
     * @param endDate
     * @constructor
     */
    public GenerateConversionReport(
        fromCurrency: string,
        toCurrency: string,
        startDate: Date,
        endDate: Date
    ): string {
        const conversions: number[] = [];

        const currentDate = new Date(startDate);

        while (currentDate <= endDate) {

            try {
                const exchangeRate: number | null = this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
                if (exchangeRate != this.EXCHANGE_RATE_ERROR) {
                    this.validateExchangeRate(exchangeRate);
                    this.calculateConversion(exchangeRate, conversions, currentDate);
                }
            }
            catch (error) {
                // @todo: Inject a log service.
                // console.log(error.message, error);
                // return `Conversion Report:\n${conversions.join('\n')}`;
                break;
            }

        }

        return `Conversion Report:\n${conversions.join('\n')}`;
    }

    /**
     * Service call for exchange rate.
     *
     * @param fromCurrency
     *   The currency to convert from.
     * @param toCurrency
     *   The currency to convert to.
     *
     * @return number|null
     *   The exchange rate or null if an error occurred.
     *
     * @private
     */
    private getExchangeRate(fromCurrency: string, toCurrency: string)    {
        try {
            return this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
        }
        catch (error) {
            if (error instanceof InvalidCurrencyError) {
                // @todo: Inject a log service.
                console.log(error.message, error);
                return this.EXCHANGE_RATE_ERROR;
            }
            console.log(error);
            throw new NetworkError('The network is down.');
        }

    }

    private calculateConversion(exchangeRate: number, conversions: number[], currentDate: Date) {
        const convertedAmount = this.FIXED_AMOUNT * exchangeRate; // Assume a fixed amount for simplicity
        conversions.push(convertedAmount);
        currentDate.setDate(currentDate.getDate() + 1);
    }

    private validateExchangeRate(exchangeRate: number) {
        if (!exchangeRate) {
            throw new Error('Unable to fetch exchange rate.');
        }

        if (isNaN(exchangeRate)) {
            throw new Error('Invalid exchange rate.');
        }
    }

    /**
     * Validate the input amount.
     *
     * @param amount number
     *   The amount to validate.
     *
     * @throws InvalidInputAmountError
     *   If the amount is invalid.
     * @throws RangeError
     *   If the amount is negative.
     *
     * @private
     */
    private validateAmount(amount: number): void {
        if (isNaN(amount)) {
            throw new InvalidInputAmountError('Invalid amount input.');
        }
        if (amount < 0) {
            throw new RangeError('The amount must be a positive number.');
        }
    }
}
