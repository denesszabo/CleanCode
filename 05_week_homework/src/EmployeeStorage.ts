import { Employee } from "./Employee";
import {EmployeeNotFound} from "./error/EmployeeNotFound";
export class EmployeeStorage {

    private employees: Employee[] = [];
    constructor() {
        this.employees.push(new Employee('John', 10000));
        this.employees.push(new Employee('Bon', 8000));
        this.employees.push(new Employee('Amanda', 11000));
    }

    public getEmployeeByName(name: string): Employee {
        for ( let i = 0; i < this.employees.length; i++) {
            if (this.employees[i].getName() === name) {
                return this.employees[i];
            }
        }
        throw new EmployeeNotFound('Employee not found');
    }

    public addEmployee(employee: Employee): void {
        this.employees.push(employee);
    }

    public getEmployees(): Employee[] {
        return this.employees;
    }

    public updateEmployee(changed: Employee): boolean {
        let success = false;
        for ( let i = 0; i < this.employees.length; i++) {
            if (this.employees[i].getName() === changed.getName()) {
                this.employees[i] = changed;
                success = true;
            }
        }

        console.log(this.employees);

        return success;
    }
}
