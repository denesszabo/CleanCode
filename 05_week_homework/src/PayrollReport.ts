import {Employee} from "./Employee";

export class PayrollReport {

    public generateReport(employees: Employee[]): string {
        let report = [];
        if (employees.length === 0) {
            return '';
        }

        employees.forEach((employee) => {
            report.push("Name: " + employee.getName() + ", salary: " + employee.calculateSalary());
        });

        report.push("Total salary: " + this.calculateTotalPayroll(employees));
        return report.join("\n");
    }

    public calculateTotalPayroll(employees: Employee[]): number {
        let totalPayroll = 0;
        employees.forEach((employee) => {
            totalPayroll += employee.calculateSalary();
        });
        return totalPayroll;
    }
}
