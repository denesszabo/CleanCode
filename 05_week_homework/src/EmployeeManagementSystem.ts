import {EmployeeStorage} from "./EmployeeStorage";
import {Employee} from "./Employee";
import {PayrollReport} from "./PayrollReport";

export class EmployeeManagementSystem {
    private employees: Employee[] = [];

    constructor(private employeeStorage: EmployeeStorage, private payrollReport: PayrollReport) {
        this.employeeStorage = employeeStorage;
    }

    public addEmployee(employee: Employee): void {
        // Real-world code to add employee to the system
        this.employeeStorage.addEmployee(employee);
    }

    public generateReports(): string {
        const employees = this.employeeStorage.getEmployees();
        const report = this.payrollReport.generateReport(employees);
        console.log(report);

        return report;
    }

    public calculatePayroll(): number {
        const employees = this.employeeStorage.getEmployees();
        const totalPayroll = this.payrollReport.calculateTotalPayroll(employees);
        return totalPayroll;
    }

    public getEmployeeByName(name: string): Employee {
        return this.employeeStorage.getEmployeeByName(name);
    }
    public promoteEmployee(employee: Employee, newSalary: number): void {
        employee.setSalary(newSalary);
        this.employeeStorage.updateEmployee(employee);
    }
}
