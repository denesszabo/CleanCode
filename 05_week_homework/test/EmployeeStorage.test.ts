import { EmployeeStorage } from '../src/EmployeeStorage';
import {Employee} from "../src/Employee";

describe('Employee tests', () => {
    it('Should return the employee John', () => {
        // Arrange
        const employeeStorage = new EmployeeStorage();

        // Act1
        const actualResult: Employee | null = employeeStorage.getEmployeeByName('John');

        // Assert
        expect(actualResult).toMatchSnapshot('John');
    })

    it('Should return all employee ', () => {
        // Arrange
        const employeeStorage = new EmployeeStorage();

        // Act
        const actualResult = employeeStorage.getEmployees();

        // Assert
        expect(actualResult).toMatchSnapshot('John');
        expect(actualResult).toMatchSnapshot( 'Bon');
        expect(actualResult).toMatchSnapshot( 'Amanda');
    })
    // @todo This test is working just for my test data!
    it('Should add one more employee to the list', () => {
        return;
        // Arrange
        const employeeStorage = new EmployeeStorage();
        const employee: Employee = new Employee('Wick', 20000)
        // Act
        employeeStorage.addEmployee(employee);

        const actualResult = employeeStorage.getEmployees();
        // Assert
        expect(actualResult.length).toBe(4);
        expect(actualResult).toMatchSnapshot('John');
        expect(actualResult).toMatchSnapshot( 'Bon');
        expect(actualResult).toMatchSnapshot( 'Amanda');
        expect(actualResult).toMatchSnapshot( 'Wick');
    })

    // @todo This test is working just for my test data!
    it('Should update an employee', () => {
        return;
        // Arrange
        const employeeStorage = new EmployeeStorage();
        const newName = 'Samantha';
        let employee: Employee = employeeStorage.getEmployeeByName('Amanda');

        // Assert
        expect(employee).not.toBeNull();

        // Act
        employee.setName(newName);
        employee.setSalary(25000);
        const success:boolean = employeeStorage.updateEmployee(employee);

        // Assert
        expect(success).toBe(true);

        // Act
        employee = employeeStorage.getEmployeeByName(newName);

        // Assert
        expect(employee.getName()).toBe('Samantha');
        expect(employee.getSalary()).toBe(2500);
    })

})
