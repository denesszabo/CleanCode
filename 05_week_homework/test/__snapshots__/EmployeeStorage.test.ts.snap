// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee tests Should return all employee : Amanda 1`] = `
[
  Employee {
    "name": "John",
    "salary": 10000,
  },
  Employee {
    "name": "Bon",
    "salary": 8000,
  },
  Employee {
    "name": "Amanda",
    "salary": 11000,
  },
]
`;

exports[`Employee tests Should return all employee : Bon 1`] = `
[
  Employee {
    "name": "John",
    "salary": 10000,
  },
  Employee {
    "name": "Bon",
    "salary": 8000,
  },
  Employee {
    "name": "Amanda",
    "salary": 11000,
  },
]
`;

exports[`Employee tests Should return all employee : John 1`] = `
[
  Employee {
    "name": "John",
    "salary": 10000,
  },
  Employee {
    "name": "Bon",
    "salary": 8000,
  },
  Employee {
    "name": "Amanda",
    "salary": 11000,
  },
]
`;

exports[`Employee tests Should return the employee John: John 1`] = `
Employee {
  "name": "John",
  "salary": 10000,
}
`;
